### ISSUE TITLE
How does fit_generator works

### ISSUE THREAD

**Issue by nameofuser1** [fchollet/keras#3809]
I'm not quite understand how fit_generator works. In particular what does mean parameter samples_per_epoch? As described [here](https://github.com/fchollet/keras/issues/1627) samples_per_epoch = batch_size * numbers of batches and fit_generator will read exactly samples_per_epoch samples from generator. And there also said that we control batch size by our generator yieldings. But here begins what I'm really missing.

Let's assume that generator return batch of 5 single samples and we want to train model on 10 batches per epoch so we have training on 50 single samples per epoch. But if we set samples_per_epochs with given formula we get 50 samples taken by fit_generator and each sample is a BATCH yielded by generator so we get 5*50=250 single samples.  I see that we can get samples number in a batch by dividing input length by length of input shape's zero axis. Therefore we can calculate how much times we need to ask for a batch. But it seems too intricately for me.

Ok. Then let's suppose that the formula is wrong and in this case we set samples_per_epoch to our batch_size. It looks more confident for me. Then generator will take exactly sample_per_epoch batches from generator. Again batch size in units of single samples can be computed by dividing.

So the question is what is actually right?
